### PR TITLE
[WFLY-18582] Add prometheus endpoint to micrometer extension

### PR DESCRIPTION
**Pull request by jasondlee**
https://issues.redhat.com/browse/WFLY-18582

Bump schema version
Start modifying WildFlyRegistry to better support composite Add child element for OTLP registry
Add child element for Prometheus registry
Add support for a Prometheus endpoint
Add module for Prometheus client
Add Prometheus simpleclient dep
Add resource transformer
Update XML description

____

More information about the [wildfly-bot[bot]](https://github.com/wildfly/wildfly-github-bot)

### CODE REVIEW

File: observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusRegistryDefinition.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.wildfly.extension.micrometer.MicrometerConfigurationConstants;
+import org.wildfly.extension.micrometer.MicrometerExtension;
+import org.wildfly.extension.micrometer.registry.WildFlyCompositeRegistry;
+
+public class PrometheusRegistryDefinition extends SimpleResourceDefinition {
+    public static final PathElement PATH_ELEMENT = PathElement.pathElement("registry", "prometheus");
+
+    static final RuntimeCapability<Void> MICROMETER_PROMETHEUS_CONTEXT_CAPABILITY =
+            RuntimeCapability.Builder.of("org.wildfly.extension.micrometer.prometheus.http-context",
+                            PrometheusContextService.class)
```
Comment (pferraro) [inline observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusRegistryDefinition.java:29]: Capability service types should _never_ leak the MSC API.
It seems to me that this capability either should not have any service type (if there are no services that would consume it) or use a service type of Void.

File: observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    static void install(OperationContext operationContext,
+                        WildFlyPrometheusRegistry registry,
+                        String prometheusContext,
+                        boolean securityEnabled) {
+        ServiceBuilder<?> serviceBuilder = operationContext.getCapabilityServiceTarget()
+                .addService(MICROMETER_PROMETHEUS_CONTEXT_CAPABILITY.getCapabilityServiceName());
+
+        Supplier<ExtensibleHttpManagement> extensibleHttpManagement =
+                serviceBuilder.requires(operationContext.getCapabilityServiceName(HTTP_EXTENSIBILITY_CAPABILITY, ExtensibleHttpManagement.class));
+        Consumer<PrometheusContextService> prometheusContextConsumer =
```
Comment (pferraro) [inline observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java:40]: A service should never provide itself unless via an interface that does not extend Service.

File: observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    private final Supplier<ExtensibleHttpManagement> extensibleHttpManagement;
+    private final Supplier<Boolean> securityEnabledSupplier;
+    private final String prometheusContext;
+    private final WildFlyPrometheusRegistry registry;
+
+    static void install(OperationContext operationContext,
+                        WildFlyPrometheusRegistry registry,
+                        String prometheusContext,
+                        boolean securityEnabled) {
+        ServiceBuilder<?> serviceBuilder = operationContext.getCapabilityServiceTarget()
```
Comment (pferraro) [inline observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java:35]:
```java
operationContext.getCapabilityServiceTarget().addCapability(MICROMETER_PROMETHEUS_CONTEXT_CAPABILITY)
```

**Review by pferraro — CHANGES_REQUESTED** [review]

File: observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java (hunk 3 above)
Comment (jasondlee) [reply observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java:35]: @pferraro Am I supposed to add that line or replace line 35 with it?

File: observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java (hunk 2 above)
Comment (jasondlee) [reply observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java:40]: I'm not sure what to do here. I patterned this off of the existing (and admittedly old) Metrics context service.

File: observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java (hunk 3 above)
Comment (pferraro) [reply observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java:35]: Replace

File: observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java (hunk 2 above)
Comment (jasondlee) [reply observability/micrometer/src/main/java/org/wildfly/extension/micrometer/prometheus/PrometheusContextService.java:40]: @pferraro Any extra guidance here?

**Comment by darranl** [comment]
I am just dropping the 32.x label from this one as we need to start following the WildFly feature development process.

File: observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerResourceDefinitionRegistrar.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                    MicrometerResourceDefinitionRegistrar.class);
+
+    MicrometerResourceDefinitionRegistrar() {
+        ResourceDescriptor.Builder builder =
+                ResourceDescriptor.builder(RESOLVER.createChildResolver(MicrometerSubsystemRegistrar.PATH));
+
+        registration = ResourceRegistration.of(MicrometerSubsystemRegistrar.PATH);
+        descriptor = builder.addCapability(MICROMETER_COLLECTOR_RUNTIME_CAPABILITY)
+                .addCapability(MICROMETER_REGISTRY_RUNTIME_CAPABILITY)
+                .build();
```
Comment (pferraro) [inline observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerResourceDefinitionRegistrar.java:91]: I would move this to your register(...) method.  These don't need to be member variables.

**Pull request closed without merging**